src/test/java/pages/BasePage.java

```
@@ -0,0 +1,138 @@
+        this.wait = new WebDriverWait(driver, Duration.ofSeconds(baseWaitInSeconds));
+        this.baseUrl = ConfigReader.getProperty(ConfigReader.PropertyKeys.BASE_URL);
+    }
+
+    public By buttonNavbarLocatorByDataId(NavbarButton button) {
+        return By.xpath(navbarXpath + "//li[@data-id='"+ button.getDataId() +"']//span[@class='menu_text']");
+    }
+
+    public By buttonCategoryMenuLocatorByText(CategoryMenuButton button) {
+        return By.xpath(categoryMenuXpath + "//a[contains(text(), '" + button.getText() + "')]");
+    }
+
+    public By hyperLinkFooterLocatorByText(FooterHyperLink link) {
+        return By.xpath(footerlinksXpath + "//a[text()='" + link.getText() + "']");
+    }
+
+    public void openUrl(String url) {
+        driver.get(url);
+    }
+
+    public void refresh() {
+        driver.navigate().refresh();
+    }
+
+    public void clickOnElement(By locator) {
+        locateElement(locator).click();
+    }
+
+    public void scrollToElement(By locator) {
+        WebElement element = locateElement(locator);
+        JavascriptExecutor js = (JavascriptExecutor) driver;
+        js.executeScript("arguments[0].scrollIntoView(true);", element);
+    }
+
+    public void hoverOverElement(By locator) {
+        WebElement element = locateElement(locator);
+        Actions actions = new Actions(driver);
+        actions.moveToElement(element).perform();
+    }
+
+    public void selectCheckbox(By locator) {
+        WebElement element = locateElement(locator);
+        if (!element.isSelected()) {
+            element.click();
+        }
+    }
+
+    public void deselectCheckbox(By locator) {
+        WebElement element = locateElement(locator);
+        if (element.isSelected()) {
+            element.click();
+        }
+    }
+
+    public void selectRadioButton(By locator) {
+        WebElement element = locateElement(locator);
+        if (!element.isSelected()) {
```

I'm always cautious when using `if`s in automated tests - I prefer my tests to be pretty 'silly', there's already too much logic in the backend/frontend implementation :)

Let's imagine:
- you would like to select a radio button
- you assume that the radio button is unselected
- but there's a bug: it is already selected 🐛 
- then our automated test happily skips the conditional and moves on with further testing 🤷‍♀️ 

Maybe we would like the test to fail if the radio button is already selected? Maybe we should throw an exception here? Worth considering :)

---

Interesting observation, I will definitely pay more attention now . I changed the methods that check the state by adding a `throw exception` like this:
```java
    public void selectRadioButton(By locator) {
        WebElement element = locateElement(locator);
        if (element.isSelected()) {
            throw new IllegalStateException("Radio button is already selected");
        }
        element.click();
    }
```